### ISSUE TITLE
Label (sub)circuits and draw visual boxes

### ISSUE THREAD

**Issue by eddieschoute** [Qiskit/qiskit#2944]
### What is the expected behavior?
I was recently constructing some circuits and thought it would be nice to see what part of the circuit implements what. To that end I think it would be nice to have labels on the circuit drawer.

Of course it is already possible to turn a subcircuit into an opaque box through a custom instruction with label, however, I'd want to be able to inspect whatever happens inside the subcircuit.

Some examples of circuit labels are:
* The curly braces and the O(sqrt(N)) of the section of Grover's algorithm that is repeated O(sqrt(N)) times in Fig 6.1 of Nielsen and Chuang.
* The dashed box, highlighting & label in the following circuit
![Quantum-teleportation-circuit-See-Figure-13-for-quantum-gate-definitions-Single](https://user-images.githubusercontent.com/25747714/62717973-e3f0f180-b9d2-11e9-863f-4d20de8b989c.png)

I'm mostly imagining something like the second example but with a possible label & not necessarily colored.

Would this be reasonably possible to fit in the circuit drawing? Especially drawing the box around the subcircuit may be tricky and the positioning of the label. Possibly also related to #1298 .

**Comment by maddy-tod** [comment]
Are you thinking of when you draw circuits like this?
```
qc1 = QuantumCircuit(2)
qc2 = QuantumCircuit(2)
qc1.h(0)
qc2.cx(0, 1)
qc1 += qc2

circuit_drawer(qc1, output='mpl')
```

I think this would be non trivial to do as currently the drawing is based off the DAG, and the DAG has no awareness that the circuit it represents is made up of multiple subcircuits. This is because when circuits are appended, they are simply added [gate by gate](https://github.com/Qiskit/qiskit-terra/blob/690e00ca7c89f8935f31e6ee1c0d08784eef54e0/qiskit/circuit/quantumcircuit.py#L207) with no storage of the fact there were originally multiple circuits. Therefore we would have to find a way to start storing this information if we wanted to use it for drawing.

**Comment by eddieschoute** [comment]
Yes, I think it would make sense to use actual circuits to identify subcircuits of the whole circuit. I have another issue up #2959 that also relates to selecting subcircuits.
Maybe instead of adding metadata and adding everything gate by gate you'd store a reference to the subcircuits instead in the DAGCircuit/Circuit class and from there it would be clear what is being referred to.

**Comment by javabster** [comment]
re-opened due to comment in #7192 indicating this issue may still be relevant 😄

**Comment by aeddins-ibm** [comment]
`box` (#13772 and related) appears to be a big step towards addressing this issue. I don't know if it's currently possible to show a label though.

<img width="528" alt="Image" src="https://github.com/user-attachments/assets/f3152718-18cf-4014-a4a6-6755d007ae8b" />

**Comment by jakelishman** [comment]
We have `box`, which can be used for at least some of this, as Andrew showed above.

Other than that, we don't really have the resources to turn the visualisers into a fully-fledged custom graphics package - I totally agree that such a thing would be useful, but pragmatically I've got to consider it out-of-scope for Qiskit.  The boxes drawn by `box` are about the limit of where we can reasonably go with this at the moment, so I'll close the issue as "complete as best as we can"..
